### DISCUSSION TITLE
Welcome to Flyfisher604/mpcnc_posts_processor Discussions!

### DISCUSSION THREAD

**Discussion by flyfisher604** [flyfisher604/mpcnc_post_processor#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by AcrimoniousMirth** [comment]
Hi!
First off, really really love the list-processor. It works a treat and is quite intuitive! Unfortunately I'm having a bit of an issue with  [Group 7: override properties](https://github.com/flyfisher604/mpcnc_post_processor#group-7-override-behaviour-by-external-file-properties). I'm really not sure what's expected here? 
I have a hunch that I'll need to create some code files in the nc folder for each operation then copy the file path and paste it into the boxes provided? Is that correct? 
As for operation, does it pick up on "Manual NC → Force Tool Change" commands and replace that with the code?
Sorry, still new and looking to get tool changes working smoothly.
Thanks for the help!

**Reply by flyfisher604** [reply]
Thanks for the comment. I'll try to review sometime in the coming week. 

Can you provide more details on what you are trying to do?

From memory I believe it simple inserts that tool change file when the tool differs between two operations. 

You shouldn't need the file path, only the name of the file. Place the file in the GCode nc output folder. 

Let me know if this helps.

**Reply by AcrimoniousMirth** [reply]
Hi,
Thanks for your reply! I did some experimenting. To get this to work these are the steps:
1. Locate the folder you have assigned to have Gcode exported to. This is the folder which you will drop your tool change (or other) script in. This folder does not need to be the default nc folder.
2. Create a new Gcode file, drop the code you want to be included in there.
3. Copy the name of the Gcode file and open F360. Navigate to the Post Processor and scroll down to the section dealing with external scripts.
4. Paste the file name in there.

The one drawback I see is that these control scripts are now mixed in amongst the Gcode exports... but otherwise this seems to work quite nicely! I included my initials in the Gcode snippet so that I can easily navigate to the snippet in the larger output file by searching for them.

**Comment by RufanMelfor** [comment]
Dear,

Thanks for the post possessor, looks very promising.
I use Duet3d with RepRap and like to use the laser mode.
I am not a programmer hence this question, can you guide me with the following?
For laser mode RepRap Fimware > 3.x Laser power (PWM) I need behind each G1 a S command.
For example G1 X10 Y100 F1000 S100

https://docs.duet3d.com/User_manual/Reference/Gcodes#m452-select-laser-device-mode

I'd love to hear your opinion
Thanks in advance.

**Comment by Breed2085** [comment]
First off thank you for putting in the hard work to make this.
I have a question... when using fusion 360 2d contour the radius compensation is default to "in computer". when I run the code with that default the part is off by the size of the bit. the only way to get the proper size is to turn off the radius compensation. Is this normal or am I doing something wrong?

**Reply by flyfisher604** [reply]
Glad you like the post processor (PP).

The PP makes no changes to the coordinates created by F360. The PP does make some G1s into G0s and it does limit/scale feedrates.

I am a hobbyist F360 user so your question may be much better answered in a F360 forum. That said, I always use In Computer or Off (for engraving). 

This article may help https://www.autodesk.com/products/fusion-360/blog/cutter-compensation/

If you do find an issue caused by the PP please let me know. 

Don.

**Comment by nickolas1969** [comment]
Getting back into my MPCNC with Fusion, and saw Fusion made a few changes. Was A Guffy PP user in the past, and found yours as first in my search for a current one. Looks great! Did you have support for Router'Laser ON/OFF commands for Marlin?  I had been using M106/M107 Fan PWM's for this on my build. 

Router: M106 P0 = ON, M107 P0 = OFF
Laser: M106 P2 Sxxx = ON to level XXX, M107 P2 = OFF.

**Reply by nickolas1969** [reply]
Whoops, I just saw the group for the Laser is there.. I missed that, and appears just as I remember Guffy's doing it. 

In teh ROuter portion, I can only see that I can enable a manual pause, but if it is not enables, just no pause, did not see where I can tell the PP how my routers turns on and off (I'll keep digging)

**Reply by flyfisher604** [reply]
Hope that section addresses your needs. Others are using this PP with Marlin without issues, as far as I know. That said, I may have missed something as I don't have a laser setup. 

Been a few years since I've made any changes to the PP's code - so adding something would require me becoming reacquainted with the code. :-) That code take some time. 

Let me know if you still have issues. The MPCNC forum is also an excellent place to ask these questions. You are likely to find someone already doing what you want to do.

**Reply by nickolas1969** [reply]
So the Laser portion is good, as Laser ON/Laser Off is something that happens as part of any laser job. A Spindle is typically just once. ON at the start of job, OFF at the end, so really, pre and post g-code headers was where I think Guffy had his. 

Short term, it just means i edit every NC Program looking for the Pause inserted by your PP, and change it to my ON/OFF commands. 

Otherwise I would need a section for pre/post command headers section, I think?

**Reply by nickolas1969** [reply]
For my specific needs, actually, it would be awesome if that Pause Command, When OFF, Would ask for a spindle ON/OFF command as an option and insert that instead.

**Reply by nickolas1969** [reply]
Ok, so hold up, I am teaching myslef PP code by looking at yours.. So Spindle ON and Off are sections after all, looks like a Marlin 2.0 version of the G3/G4 format could be added.


```
function spindleOn(_spindleSpeed, _clockwise) {
  // Is Grbl?
  if (fw == eFirmware.GRBL) {
    writeComment(eComment.Important, " >>> Spindle Speed " + speedFormat.format(_spindleSpeed));
    writeBlock(mFormat.format(_clockwise ? 3 : 4), sOutput.format(spindleSpeed));
  }

  // Default
  else {
    if (properties.job2_ManualSpindlePowerControl) {
      // For manual any positive input speed assumed as enabled. so it's just a flag
      if (!this.spindleEnabled) {
        writeComment(eComment.Important, " >>> Spindle Speed: Manual");
        askUser("Turn ON " + speedFormat.format(_spindleSpeed) + "RPM", "Spindle", false);
      }
    } else {
      writeComment(eComment.Important, " >>> Spindle Speed " + speedFormat.format(_spindleSpeed));
      writeBlock(mFormat.format(_clockwise ? 3 : 4), sOutput.format(spindleSpeed));
    }
    this.spindleEnabled = true;
  }
}
```

**Reply by nickolas1969** [reply]
BTW: Thank you for this neat PP example, I'm learning Fusion PP language.. erm.. I think. I should be able to hack this enough to insert an if clause for Marlin over the default. I just need to see a section how a string is set. 

Add a section in Group 1:

Job: Manual Spindle On/Off: (check box)
Auto Spindle On: (accept g-code string)
Auto Spindle Off: (accept g-code string)

And just not use the Auto if Manual is checked.

**Reply by nickolas1969** [reply]
I plan on trying my mod this weekend, but for today, I created two small TXT files and added to Group 7 

; *** START begin ***
; --- Start custom gcode C:\Users\nicko\OneDrive\Documents\Fusion 360\NC Programs\RouterStart.txt
; Start Router and VAC on Fan 0 of Marlin
M106 P0 S255
; *** START end ***
<snip>
; --- Start custom gcode C:\Users\nicko\OneDrive\Documents\Fusion 360\NC Programs\RouterStop.txt
; Stop Router and Vac on Fan 0 Marlin
M107 P0

**Reply by flyfisher604** [reply]
Glad you are making progress. 

In your previous post you referred to G3/4, aren't these arc commands? M3/4 are spindle control. 

The spindle start is already outputting an M3 or M4, depending on the rotation direction.

Maybe you can try again describing the feature you would like to see.

**Reply by nickolas1969** [reply]
Sorry, was milling, can't leave that beastie unattended O.o

So all I was thinking was an augment to the checkbox feature in group 1.
If checked, it inserts the manual PAUSE command to be cleared by the user at the LCD.
If not checked, it would instead insert what is typed in the Spindle ON and Spindle Off commands added below that line.

I believe what I saw in another post processor was a check box for M106/M107 and a pick list of P values since that was the most common way to trigger a TTL Relay was to rob one of the FAN's, just like you do for Laser. 

P0 - P2 were available on most of the RepRap/Arduino line, and I used P0 (FAN0) and P2 (FAN2) on my RAMBo. 

So perhaps instead of a free form one line command, the code could be pick list bound to just common use cases like the Laser is. 

I am thinking of all the ideas folks have used since I started my journey in 2018.

**Comment by robwaldhauser** [comment]
Absolutely loving the post processor for Klipper! Question for you though, is there any way to get Estimated Time to Completion out of Fusion into Klipper? Thanks!

**Reply by flyfisher604** [reply]
The post processor only transforms F360's internal information into GCode text. F360 does have a data structure with a bunch of information in it (not sure duration is there), but I would also need to know the GCode to output to load that into your machine.
